ownCloud/Client/Actions/Actions+Extensions/OpenInAction.swift

```diff
@@ -22,61 +22,95 @@ class OpenInAction: Action {
-
-		let controller = DownloadFileProgressHUDViewController()
+		downloadProgressController = DownloadFileProgressHUDViewController()
 
-		controller.present(on: viewController) {
-			if let progress = self.core?.downloadItem(item, options: [ .returnImmediatelyIfOfflineOrUnavailable : true ], resultHandler: { (error, _, _, file) in
-				if error != nil {
-					Log.log("Error \(String(describing: error)) downloading \(String(describing: item.path)) in openIn function")
+		downloadProgressController?.present(on: viewController) { [weak self] in
 
-					self.completionHandler = { error in
+			let downloadGroup = DispatchGroup()
 
-						let appName = OCAppIdentity.shared.appName ?? "ownCloud"
-						let alertController = UIAlertController(with: "Cannot connect to ".localized + appName, message: appName + " couldn't download this file".localized, okLabel: "OK".localized, action: nil)
-						viewController.present(alertController, animated: true)
+			if let items = self?.context.items {
+				for item in items {
+					downloadGroup.enter()
+					if let progress = self?.core?.downloadItem(item, options: [ .returnImmediatelyIfOfflineOrUnavailable : true ], resultHandler: { (error, _, _, file) in
+						if error != nil {
+							Log.log("Error \(String(describing: error)) downloading \(String(describing: item.path)) in openIn function")
+							self?.downloadError = error
+						} else {
+							self?.downloadedFiles.append(file!)
+						}
+						downloadGroup.leave()
+					}) {
+						self?.downloadProgressController?.attach(progress: progress)
+						self?.publish(progress: progress)
 					}
 
-					controller.dismiss(animated: true, completion: {
-						self.completed(with: error)
-					})
-				} else {
-					OnMainThread {
-						controller.dismiss(animated: true, completion: {
-							if let fileURL = file?.url {
-								self.interactionController = UIDocumentInteractionController(url: fileURL)
-								self.interactionController?.delegate = self
-								self.interactionController?.presentOptionsMenu(from: .zero, in: viewController.view, animated: true)
-							}
-						})
+					if self?.downloadError != nil {
+						break
 					}
 				}
-			}) {
-				controller.attach(progress: progress)
-				self.publish(progress: progress)
 			}
+
+			downloadGroup.notify(queue: .main) { [weak self] in
+				self?.downloadProgressController?.dismiss(animated: true, completion: {
+					if let error = self?.downloadError {
+						self?.showDownloadError()
+						self?.completed(with: error)
+					} else {
+						self?.presentSharingViewController()
+						self?.completed()
+					}
+				})
+			}
+		}
+	}
+
+	fileprivate func showDownloadError() {
+
+		guard let viewController = context.viewController else { return }
+
+		let appName = OCAppIdentity.shared.appName ?? "ownCloud"
+		let alertController = UIAlertController(with: "Cannot connect to ".localized + appName, message: appName + " couldn't download file(s)".localized, okLabel: "OK".localized, action: nil)
```

Comment:
Do you have the context, if you are downloading one or multiple files? If yes, you could use a message for a single file (like before) or if there are multiple files, you can present the current message.

Reply:
@hosy I could show file name at this point in case it is just one file, but I dont know if it is that useable.